geoparquet/index.qmd

```diff
@@ -0,0 +1,86 @@
+
+Some of the sections below will discuss strengths of Parquet in general. Keep in mind that because GeoParquet is built on top of Parquet, GeoParquet inherits all of these strengths.
+
+### File layout
+
+Parquet files are laid out differently than other tabular formats like CSV or FlatGeobuf, so it's helpful to see a diagram:
+
+![Schematic of Parquet file layout](../images/geoparquet_layout.png)
+
+A Parquet _file_ consists of a sequence of chunks called _row groups_. These are logical groups of columns with the same number of rows. A row group consists of multiple columns, each of which is called a _column chunk_. These are sequences of raw column values that are guaranteed to be contiguous in the file. All row groups in the file must have the same schema, meaning that the data type of each column must be the same for every row group.
+
+A Parquet file includes metadata describing the internal chunking. This metadata includes the byte range of every column chunk in the dataset. This allows a Parquet reader to fetch _any given column chunk_ once they have the file metadata.
+
+The Parquet metadata also includes _column statistics_ (the minimum and maximum value) for each column chunk. This means that if a user is interested in data where column "A" has values greater than 100, the Parquet reader can skip loading and parsing any column chunks where the maximum is known to be less than 100.
+
+In Parquet, the metadata is located at the _end_ of the file rather than at the beginning. This makes it much easier to write, as you don't need to know how many total rows you have at the beginning, but makes it slightly harder to read. In practice, this is not too much more difficult to read: a Parquet reader first reads the end of the file, then makes reads for select columns.
+
+### Column oriented
+
+The bytes of each _column_ are contiguous, instead of each _row_. This means that it's easy to filter on columns — fetching _all rows of a single column_ — but not possible to filter on individual rows.
+
+### Column filtering
+
+Because Parquet is column oriented, a Parquet reader can fetch only specific columns that the user is interested in.
+
+### Row group filtering
+
+Because Parquet is internally chunked, Parquet can fetch only specific row groups that meet a specific filtering condition.
+
+Note that row group filtering on a specific column tends to only work well if the Parquet file was sorted on that column when saved. Non-sorted columns tend to have random values, and so the column statistics won't tend to filter out many row groups.
```

Comment:
If I understand this correctly, does this mean that you can really only optimize filtering row groups by one column?

Reply:
Yeah, pretty much. I think that's the biggest difference between file formats and databases. Databases can have an index on whatever column you want, and then when you run a query, it'll use all of the indexes. But that's why it's hard to make databases work as cloud-native files, because if you have high latency, you don't want to make lots of tiny fetches.

FlatGeobuf optimizes filtering by a spatial query (when an index is generated) because it sorts all features in the file by their hilbert value. Parquet lets you sort by any column you choose. You _can_ sort by multiple columns, but the effectiveness of the sort drops off substantially after the first column.

Reply:
Do you think I should try to be more specific here? Or is this a general "cloud-native" note to go on a different page?

Reply:
Good question. I think for now I would add

```
::: {.callout-note}
you can really only optimize filtering row groups by one column. This is the biggest difference between file formats and databases. Databases can have multiple indexes on whatever columns you want, and then when you run a query, and it will use all of the indexes. But that's why it's hard to make databases work as cloud-native files, because if you have high latency, you don't want to make lots of tiny fetches.
```

as a note. And then if we have a general cloud-native concepts page which explains more, we can link to that more general page.

I do think a page on databases vs files would be good, as I mentioned before, using some of this content and content from your DS presentation.